crypto/hpke/hpke_util.c

```diff
@@ -173,3 +411,106 @@ EVP_KDF_CTX *ossl_kdf_ctx_create(const char *kdfname, const char *mdname,
     }
     return kctx;
 }
+
+/*
+ * @brief look for a label into the synonym tables, and return its id
+ * @param st is the string value
+ * @param synp is the synonyms labels array
+ * @param arrsize is the previous array size
+ * @return 0 when not found, else the matching item id.
+ */
+static uint16_t synonyms_name2id(const char *st, const synonymttab_t *synp,
+                                size_t arrsize)
+{
+    size_t i, j;
+
+    for (i = 0; i < arrsize; ++i) {
+        for (j = 0; j < OSSL_NELEM(synp[i].synonyms); ++j) {
+            if (OPENSSL_strcasecmp(st, synp[i].synonyms[j]) == 0)
+                return synp[i].id;
+        }
+    }
+    return 0;
+}
+
+/*
+ * @brief map a string to a HPKE suite based on synonym tables
+ * @param str is the string value
+ * @param suite is the resulting suite
+ * @return 1 for success, otherwise failure
+ */
+int ossl_hpke_str2suite(const char *suitestr, OSSL_HPKE_SUITE *suite)
+{
+    uint16_t kem = 0, kdf = 0, aead = 0;
+    char *st = NULL, *instrcp = NULL;
+    size_t inplen;
+    int labels = 0, result = 0;
+    int delim_count = 0;
+    const char OSSL_HPKE_STR_DELIMSTR[] = { OSSL_HPKE_STR_DELIMCHAR, '\0' };
+
+    if (suitestr == NULL || suitestr[0] == 0x00 || suite == NULL) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_PASSED_NULL_PARAMETER);
+        return 0;
+    }
+    /* somebody might someday extend in an untested way... */
+    if (OPENSSL_strnlen(OSSL_HPKE_STR_DELIMSTR, 1) != 1) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        return 0;
+    }
+
+    inplen = OPENSSL_strnlen(suitestr, OSSL_HPKE_MAX_SUITESTR);
+    if (inplen >= OSSL_HPKE_MAX_SUITESTR)
+        return 0;
+    /*
+     * we don't want a delimiter at the end of the string;
+     * strtok() doesn't care about that, so we should
+     */
+    if (suitestr[inplen - 1] == OSSL_HPKE_STR_DELIMCHAR)
+        return 0;
+    /* We want exactly two delimiters in the input string */
+    for (st = (char *)suitestr; *st != '\0'; st++) {
+        if (*st == OSSL_HPKE_STR_DELIMCHAR)
+            delim_count++;
+    }
+    if (delim_count != 2)
+        return 0;
+
+    /* Duplicate `suitestr` to allow its parsing  */
+    instrcp = OPENSSL_memdup(suitestr, inplen + 1);
+    if (instrcp == NULL)
+        goto fail;
+
+    /* See if it contains a mix of our strings and numbers */
+    st = strtok(instrcp, OSSL_HPKE_STR_DELIMSTR);
+    if (st == NULL)
+        goto fail;
+
+    while (st != NULL && labels < 3) {
+        /* check if string is known or number and if so handle appropriately */
+        if (labels == 0
+            && (kem = synonyms_name2id(st, kemstrtab,
+                                      OSSL_NELEM(kemstrtab))) == 0)
```

Comment:
Off-by-one indentation error here and below (twice)

Reply:
ack

Reply:
fixed